### PR TITLE
Fonts-weight descriptor

### LINKED ISSUE #10910: @font-face/font-weight: expand on <font-weight-absolute>{2} importance
MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-weight

#### What information was incorrect, unhelpful, or incomplete?

There should be more ample space devoted to `font-weight` ranges (`<font-weight-absolute>{2}`) required for variable fonts to work reliably across browsers.

I've only just started writing my first variable font CSS today, off the cuff, so I don't have an insight on the history of the spec / known browser quirks, but pretty soon I discovered that not mentioning a `font-weight` range in `@font-face` causes issues in Chrome and Safari. 

In particular, the example in the _Variable fonts_ section of [the more famous `font-weight` CSS property of the two](https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight#variable_fonts) currently doesn't work in Chrome due to this omission. I have opened a PR on the css-examples repo to address it:

- https://github.com/mdn/css-examples/pull/56

#### What did you expect to see?

The expansion of the "Variable fonts" section with a note about the importance of adding a `font-weight` range, and ideally an sample @font-face declaration for variable fonts in the _Examples_ section.

#### Did you test this? If so, how?

Tested before/after code in the aforementioned PR in Chrome.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/css/@font-face/font-weight`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-weight
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/css/@font-face/font-weight/index.md
* Last commit: https://github.com/mdn/content/commit/26f50bdaeb1228f458d21dbdf7dc110387b20daa
* Document last modified: 2021-08-12T23:59:09.000Z

</details>

### LINKED ISSUE #32619: font-weight page doesn't explain whether it is possible to set a range, or if so, how
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-weight

### What specific section or headline is this issue about?

It's an overall issue

### What information was incorrect, unhelpful, or incomplete?

1. It doesn't explain whether it is possible for users to set a range. For example, users who have trouble reading thin text may want to set a minimum font-weight, without forcing all text to the same weight.
2. If that is possible, it doesn't explain how.
3. It mentions "multiple values" but doesn't explain if these set a range, and doesn't explain what else they do.
4. If multiple values are *supposed* to set a range, then in Firefox they are ignored as an error.

### What did you expect to see?

Information for users to set font-weight so that more pages will be readable.

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

_No response_

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/css/@font-face/font-weight`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-weight
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/css/@font-face/font-weight/index.md
* Last commit: https://github.com/mdn/content/commit/acfe8c9f1f4145f77653a2bc64a9744b001358dc
* Document last modified: 2023-07-07T07:19:19.000Z

</details>

### PR DESCRIPTION
**Pull request by estelle**
Replaces #32853
Fixes https://github.com/mdn/content/issues/32619
Fixes https://github.com/mdn/content/issues/10910

### CODE REVIEW

File: files/en-us/web/css/@font-face/font-weight/index.md [hunk 1]
```diff
@@ -80,20 +84,176 @@ People experiencing low vision conditions may have difficulty reading text set w
...
+p.two {
+  font-weight: bold;
+}
+```
+
+#### Result
+
+{{embedlivesample("Selecting normal and bold fonts", "", 300)}}
+
+### Setting font-weight ranges
```
Comment (wbamberg) [inline files/en-us/web/css/@font-face/font-weight/index.md:134]: I don't understand the point of this example. Why would someone want to use a range with a non-variable font?

ISTR when I looked into this the only use case I could see for a range was to limit the weights available in a variable range font.

Comment (estelle) [reply files/en-us/web/css/@font-face/font-weight/index.md:134]: i've actually seen this done in the wild. <s>redacted</s> had 6 different font names for the same font in different ranges, and they updated the font-family name for every font-weight property. People do it, so we need to demonstrate it.

Comment (wbamberg) [reply files/en-us/web/css/@font-face/font-weight/index.md:134]: But what's the use case for it? As someone trying to learn this feature, I look at this and think, why would anyone want to do this, what does it accomplish?

> People do it, so we need to demonstrate it.

I'm not sure I agree with this. People do all kinds of things, but we should demo sensible use cases.

Comment (wbamberg) [reply files/en-us/web/css/@font-face/font-weight/index.md:134]: Also, this:

> redacted had 6 different font names for the same font in different ranges, and they updated the font-family name for every font-weight property. 

...isn't what's happening here, is it? This has the same font-family name.

Comment (estelle) [reply files/en-us/web/css/@font-face/font-weight/index.md:134]: no; here we are using one font family name. I'll add to the pre-explanation.

File: files/en-us/web/css/@font-face/font-weight/index.md [hunk 2]
```diff
@@ -80,20 +84,176 @@ People experiencing low vision conditions may have difficulty reading text set w
...
+
+In this example, we include the same two fonts as in the example above, but we set the `font-weight` descriptors to a different ranges.
```
Comment (estelle) [inline files/en-us/web/css/@font-face/font-weight/index.md:136]:
```suggestion
In this example, we include the same two fonts as in the example above, but we set the `font-weight` descriptors to a different ranges. 

While this example is a bit contrived, it demonstrates how authors can include multiple fonts for multiple font-weights (and font-styles), by including multiple `@font-face` declarations with the same `font-family` value. In the rest of your stylesheets, you can then declare a `font-weight` (or `font-style`), and know the appropriate font will be used.
```

File: files/en-us/web/css/@font-face/font-weight/index.md [hunk 3]
```diff
@@ -80,20 +84,178 @@ People experiencing low vision conditions may have difficulty reading text set w
...
+
+While this example is a bit contrived, it demonstrates how authors can include multiple fonts for multiple font-weights (and font-styles), by including multiple `@font-face` declarations with the same `font-family` value. In the rest of your stylesheets, you can then declare a `font-weight` (or `font-style`), and know the appropriate font will be used.
```
Comment (wbamberg) [inline files/en-us/web/css/@font-face/font-weight/index.md:138]: Sorry if I'm being thick, but the previous examples demonstrates this use case as well, doesn't it? (and is not contrived, AFAIK it is showing the main use case for `font-weight`). So what is the point here of using a range?

Comment (wbamberg) [reply files/en-us/web/css/@font-face/font-weight/index.md:138]: OK, I think that makes sense! But then I still don't understand why the second range is a subset of the first. Why do we want to select the non-bold font when the dev writes `1000`?

I also wondered if you could do this without ranges, like this:

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: 1;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Light.woff2");
}
@font-face {
  font-family: "Fira Sans";
  font-weight: 200;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 400;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 900;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-ExtraBold.woff2");
}
```

What do ranges give you over this?

Comment (wbamberg) [reply files/en-us/web/css/@font-face/font-weight/index.md:138]: It looks like this^ gives some results that I would not (naively) expect, in particular `font-weight: 600;
` selected extrabold.

You do need something like this:

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: 1;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Light.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 200;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 400;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 900 1000;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-ExtraBold.woff2");
}
```

...to get the results I was expecting. I don't know why though.

Comment (wbamberg) [reply files/en-us/web/css/@font-face/font-weight/index.md:138]: It looks like this^ gives some results that I would not (naively) expect, in particular `font-weight: 600;
` selected extrabold.

With something like this:

```css
@font-face {
  font-family: "Fira Sans";
  font-weight: 1 199;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Light.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 200 399;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Regular.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 400 899;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-Bold.woff2");
}

@font-face {
  font-family: "Fira Sans";
  font-weight: 900;
  src: url("https://mdn.github.io/shared-assets/fonts/FiraSans-ExtraBold.woff2");
}
```

... I get the results I was expecting. I don't know why though.

Comment (estelle) [reply files/en-us/web/css/@font-face/font-weight/index.md:138]: i used something similar to your example; the ranges are slightly broader so i can explain the cascade (and so we don't need to clarify 399.5)

Comment (wbamberg) [reply files/en-us/web/css/@font-face/font-weight/index.md:138]: Thanks! This makes sense to me. I think I would still prefer it without the cascade demo (i.e. presenting each range "in order"), but this is OK for me.

File: files/en-us/web/css/@font-face/font-weight/index.md [hunk 4]
```diff
@@ -80,20 +84,206 @@ People experiencing low vision conditions may have difficulty reading text set w
...
+
+### Setting a range for a variable font
+
+In this example, you can change the font size of the variable font. We include ["League Mono"](https://www.theleagueofmoveabletype.com/league-mono), using a single `@font-face` at-rule. We've used a `font-weight: 300 700` value which effectively limits the range of weights that are available.
+When the `font-weight` property is specifies a weight outside this range the weight it gets is clamped to the range.
+
+#### HTML
+
+We include a paragraph with `<output>` initially set to `400`, as that is the the default font-weight for unstyled paragraph text. This paragraph is nestled between two other paragraphs to enable comparing rendered versus declared font weight values.
```
Comment (wbamberg) [inline files/en-us/web/css/@font-face/font-weight/index.md:226]:
```suggestion
We include a paragraph with `<output>` initially set to `400`, as that is the the default font-weight for unstyled paragraph text. This paragraph is nestled between two other paragraphs, so you can compare rendered versus declared font weight values.
```

**Review by wbamberg — COMMENTED** [review]
Thanks Estelle! I like this example much better now, here's a detail-level review.

File: files/en-us/web/css/@font-face/font-weight/index.md [hunk 5]
```diff
@@ -7,11 +7,13 @@ browser-compat: css.at-rules.font-face.font-weight
...
+Unless it contains a [variable font](/en-US/docs/Web/CSS/CSS_fonts/Variable_fonts_guide), a single font file contains characters from a font family in a specific weight and style: for example, "Helvetica bold italic". To declare the font to be used for a range of font weights, declare a space-separated pair of font-weight values as the value for the `font-weight` descriptor.
 
-There are generally limited weights available for a particular font family. When a specified weight doesn't exist, a nearby weight is used. Fonts lacking bold typeface are often synthesized by the user agent. To prevent this, use the {{cssxref('font-synthesis')}} shorthand property.
+Typically, a developer will want to use fonts from a single font family in a range of different weights. To accomplish this, you can define multiple {{cssxref("@font-face")}} at-rules for the same family (all with the same {{cssxref("@font-face/font-family", "font-family")}} descriptor value), one for each weight, and set the `font-weight` descriptor to define the weight range for which that particular font file will be used.
```
[the remark this answers is not in the document]
Comment (estelle) [reply files/en-us/web/css/@font-face/font-weight/index.md]: Thanks for the review. I rewrote the intro so i think its a bit better than my mishmash i had before.

**Comment by wbamberg** [comment]
I like this a lot better but I can't re-review until the live samples in previews are fixed, which I think it a thing neither of us can do.

**Comment by estelle** [comment]
@wbamberg  They are showing up for me locally. The console on this PR is reporting a 404 and not a cors issue, so I think the missing files are a PR environment issue, and merging should be ok.

> I like this a lot better but I can't re-review until the live samples in previews are fixed, which I think it a thing neither of us can do.

**Comment by estelle** [comment]
Screenshot 2024-11-25 at 4 07 05 PM

File: files/en-us/web/css/@font-face/font-weight/index.md [hunk 6]
```diff
@@ -80,20 +87,214 @@ People experiencing low vision conditions may have difficulty reading text set w
...
+}
+
+p.nine {
+  font-weight: 900;
+}
+```
+
+#### Result
+
+{{embedlivesample("Setting font-weight ranges", "", 400)}}
```
Comment (wbamberg) [inline files/en-us/web/css/@font-face/font-weight/index.md:223]:
```suggestion
{{embedlivesample("Setting font-weight ranges", "", 500)}}
```

**Review by wbamberg — APPROVED** [review]
👍 thanks @estelle !

**Pull request merged**